### PR TITLE
feature(whl): add rlhf pipeline.

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/bonus/ppof.py [hunk 1]
```diff
@@ -18,6 +19,7 @@
 from .model import PPOFModel
 from .config import get_instance_config, get_instance_env, get_hybrid_shape
 from ding.bonus.common import TrainingReturn, EvalReturn
+from ..framework.middleware.collector import ChatCollector
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:22]: merge it into `ding.framework`

File: ding/framework/middleware/collector.py [hunk 2]
```diff
@@ -190,4 +191,64 @@ def __call__(self, ctx: "OnlineRLContext") -> None:
                 break
 
 
+class ChatCollector:
+    """
+    Overview:
+        The class of the collector running by steps, including model inference and transition \
+            process. Use the `__call__` method to execute the whole collection process.
```
Comment (PaParaZz1) [inline ding/framework/middleware/collector.py:198]: why indent here

File: ding/model/common/utils.py [hunk 3]
```diff
@@ -21,3 +21,20 @@ def create_model(cfg: EasyDict) -> torch.nn.Module:
     import_module(cfg.pop('import_names', []))
     # here we must use the pop opeartion to ensure compatibility
     return MODEL_REGISTRY.build(cfg.pop("type"), **cfg)
+
+
+def top_p_logits(logits, topp=0.9, filter_value=0, min_topk=1):
+    """
+    Filter a distribution of logits using nucleus (top-p) filtering
```
Comment (PaParaZz1) [inline ding/model/common/utils.py:28]: polish comments add add unittest

File: ding/model/common/utils.py [hunk 4]
```diff
@@ -21,3 +21,20 @@ def create_model(cfg: EasyDict) -> torch.nn.Module:
...
+    https://github.com/OpenLMLab/MOSS/blob/e088f438d1a95d424c6dffef0d73134ebe62cb72/models_jittor/generation.py#L146
+    """
+    cum_logits = logits.clone()
+    if topp > 0:
+        logits_sorted, inds = torch.sort(logits, dim=-1, descending=True)
+        mask = (logits_sorted.cumsum(dim=-1) - logits_sorted) >= topp
+        mask[:, :min_topk] = False
```
Comment (PaParaZz1) [inline ding/model/common/utils.py:35]: `..., :min_topk`

File: ding/model/template/vac.py [hunk 5]
```diff
@@ -1,4 +1,7 @@
 from typing import Union, Dict, Optional
+
```
Comment (PaParaZz1) [inline ding/model/template/vac.py:2]: move these modifications to a new single file: `lm_vac.py`

File: ding/reward_model/language_reward_model.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import torch
+from transformers.models.llama.modeling_llama import LlamaForCausalLM
```
Comment (PaParaZz1) [inline ding/reward_model/language_reward_model.py:2]: add try-except import like previous implementations using transformers

File: dizoo/chat/env.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            reward_model_path,
+            tokenizer_path,
+            data_path,
+            maxlen_prompt,
+            maxlen_res,
+    ):
+        self.batch_size = batch_size
+        self.tokenizer = get_tokenizer(tokenizer_path)
+        self.rm = LlamaRewardModel.from_pretrained(reward_model_path, tokenizer=self.tokenizer, opt=None)
+        self.action_space = None
```
Comment (PaParaZz1) [inline dizoo/chat/env.py:21]: why None here

File: dizoo/chat/env.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    def reset(self):
+        self.last_batch = next(self.generator)
+        self._init_flag = True
+        return self.last_batch
+
+    def __repr__(self) -> str:
+        return "DI-engine Chat Env"
+
+    def seed(self, seed):
+        self._seed = 0
```
Comment (PaParaZz1) [inline dizoo/chat/env.py:51]: `self._seed = seed`

File: dizoo/chat/env.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    def clone(self, caller):
+        return self
```
Comment (PaParaZz1) [inline dizoo/chat/env.py:54]: deepcopy

File: ding/reward_model/language_reward_model.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class LlamaRewardModel(LlamaForCausalLM):
+
+    def __init__(self, config, opt, tokenizer):
+        super().__init__(config)
+        self.opt = opt
```
Comment (PaParaZz1) [inline ding/reward_model/language_reward_model.py:9]: why define opt here

File: ding/reward_model/language_reward_model.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        output = self.model.forward(
+            input_ids=decoder_input, attention_mask=attention_mask, return_dict=True, use_cache=False
+        )
+
+        if only_last:
+            logits = self.reward_head(output.last_hidden_state[:, -1, :]).squeeze(-1)
+        else:
+            logits = self.reward_head(output.last_hidden_state).squeeze(-1)
+
+        return (logits, )
```
Comment (PaParaZz1) [inline ding/reward_model/language_reward_model.py:24]: why return a tuple here

File: dizoo/chat/env.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            mode='train',
+        )
+        self.generator = self.dataset.final_generator()
+        self.last_batch = None
+
+    def close(self) -> None:
+        self._init_flag = False
+
+    def reset(self):
+        self.last_batch = next(self.generator)
```
Comment (PaParaZz1) [inline dizoo/chat/env.py:43]: Do you need to restrat generatore here?

File: ding/reward_model/language_reward_model.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import torch
+from transformers.models.llama.modeling_llama import LlamaForCausalLM
+
+
+class LlamaRewardModel(LlamaForCausalLM):
+
+    def __init__(self, config, opt, tokenizer):
```
Comment (PaParaZz1) [inline ding/reward_model/language_reward_model.py:7]: Should we move the creation of tokenizer insides the constructor of RM?

File: launch_ppof.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline launch_ppof.py:1]: move it to `dizoo/chat/entry`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]
